ios/viewmodels/view-model-reuse.md

```diff
@@ -0,0 +1,167 @@
+    case 0:
+      return TitleMessageCellViewModel(title: "First Name", message: profile.firstName)
+    case 1:
+      return TitleMessageCellViewModel(title: "Surname", message: profile.lastName)
+    ...
+  }
+}
+
+```
+
+This implementation abstracts out the view logic details out of the table view cell, albeit moving it to the view controller. When the view logic is trivial, this is not too bad.
+
+The implementation for ReviewViewController, however, might look like this:
+
+```swift
+
+// ReviewViewController.swift
+
+func viewModel(forIndex index: Int) {
+  switch index {
+    case 0:
+      let formatter = PhoneNumberFormatter(forCountryCode: currentCountryCode)
+      let number = formatter.formatNumber(profile.phoneNumber) ?? "-"
+      return TitleMessageCellViewModel(title: "Contact number", message: number, showsEditButton: profile.canEdit)
+    case 1:
+      let rows = topics.filter({ $0.selected }).map({ $0.title }).joinWithSeparator("\n")
+      return TitleMessageCellViewModel(title: "Visit Details", message: topics, showsEditButton: review.canEdit)
+    ...
+  }
+}
+
+```
+
+The problem with this approach is that when the view logic is trivial (ProfileViewController), we can get away with implementing view logic within the view controller, but as soon as it gets complicated(ReviewViewController), the view controller starts knowing a lot about view logic details, and we are back to view controller and view bloat.
+
+
+### Protocol based view model
+
+We can use protocols to standardize the interface for the view, while keeping implementation details abstracted out into it's own entity.
+
+```swift
+
+protocol TitleMessageCellConforming {
+  let title: String { get }
+  let message: String { get }
+  let showsEditButton: Bool { get }
+}
+
+```
+
+Note that the only difference between the protocol and the concrete implementation of this class is the type. With this, the details of each cell in ReviewViewController can be abstracted out.
+
+```swift
+
+// PhoneNumberTitleMessageCellViewModel.swift
+struct PhoneNumberTitleMessageCellViewModel: TitleMessageCellConforming {
+  let title: String
+  let message: String
+  let showsEditButton: Bool
+
+  init(profile: Profile) {
+    let formatter = PhoneNumberFormatter(forCountryCode: currentCountryCode)
+    let number = formatter.formatNumber(profile.phoneNumber) ?? "-"
+    title = "Contact number"
+    message = number
+    showsEditButton = profile.canEdit
+  }
+
+// TopicsTitleMessageCellViewModel.swift
+struct TopicsTitleMessageCellViewModel: TitleMessageCellConforming {
+  let title: String
+  let message: String
+  let showsEditButton: Bool
+
+  init(topics: [Topics], review: Review) {
+    title = "Visit Details"
+    message = topics.filter({ $0.selected }).map({ $0.title }).joinWithSeparator("\n")
+    showsEditButton = review.canEdit
+  }
+}
+
+```
+
+This moves view logic code into the view models, leaving a very nice and clean implementation for our view controller:
```

Comment:
Sorry no, I'm saying the formatting can be extracted into a VM without using protocol based VMs.  (It would just go into the main VM, not the cell VM.)